amaranth/lib/fixed.py

```diff
@@ -0,0 +1,251 @@
+from ..hdl import ast
+from .._utils import bits_for
+
+
+__all__ = ["Shape", "SQ", "UQ", "Value", "Const"]
+
+
+class Shape(ast.ShapeCastable):
+    def __init__(self, i_or_f_width, f_width = None, /, *, signed):
+        if f_width is None:
+            self.i_width, self.f_width = 0, i_or_f_width
+        else:
+            self.i_width, self.f_width = i_or_f_width, f_width
+
+        self.signed = bool(signed)
+
+    @staticmethod
+    def cast(shape, f_width=0):
+        if not isinstance(shape, ast.Shape):
+            raise TypeError(f"Object {shape!r} cannot be converted to a fixed.Shape")
+
+        # i_width is what's left after subtracting f_width and sign bit, but can't be negative.
+        i_width = max(0, shape.width - shape.signed - f_width)
+
+        return Shape(i_width, f_width, signed = shape.signed)
+
+    def as_shape(self):
+        return ast.Shape(self.signed + self.i_width + self.f_width, self.signed)
+
+    def __call__(self, target):
+        return Value(self, target)
+
+    def const(self, value):
+        if value is None:
+            value = 0
+
+        return Const(value, self)
+
+    def __repr__(self):
+        return f"fixed.Shape({self.i_width}, {self.f_width}, signed={self.signed})"
+
+
+class SQ(Shape):
+    def __init__(self, *args):
+        super().__init__(*args, signed = True)
+
+
+class UQ(Shape):
+    def __init__(self, *args):
+        super().__init__(*args, signed = False)
+
+
+class Value(ast.ValueCastable):
+    def __init__(self, shape, target):
+        self._shape = shape
+        self._target = target
+
+    @staticmethod
+    def cast(value, f_width=0):
+        return Shape.cast(value.shape(), f_width)(value)
+
+    def round(self, f_width=0):
+        # If we're increasing precision, extend with more fractional bits.
+        if f_width > self.f_width:
+            return Shape(self.i_width, f_width, signed = self.signed)(ast.Cat(ast.Const(0, f_width - self.f_width), self.as_value()))
+        
+        # If we're reducing precision, truncate bits and add the top truncated bits for rounding.
+        elif f_width < self.f_width:
+            return Shape(self.i_width, f_width, signed = self.signed)(self.as_value()[self.f_width - f_width:] + self.as_value()[self.f_width - f_width - 1])
+
+        return self
+
+    @property
+    def i_width(self):
+        return self._shape.i_width
+
+    @property
+    def f_width(self):
+        return self._shape.f_width
+
+    @property
+    def signed(self):
+        return self._shape.signed
+
+    @ast.ValueCastable.lowermethod
+    def as_value(self):
```

Comment:
I think this should be:

```python
    @ast.ValueCastable.lowermethod
    def as_value(self):
        if self.signed:
            return self._target.as_signed()
        return self._target
```

or similar, otherwise multiplying two signed fixed point numbers results in an unsigned fixed point number

Reply:
This approach does not work, it seems to break initialization of wires but I'm not quite sure how

Reply:
> otherwise multiplying two signed fixed point numbers results in an unsigned fixed point number

Why would that be the case?

Reply:
I dug into this a bit more; I believe (with my limited understanding of the type system here), is that the signdedness of _target is given by the signdedness of the underlying raw value that the number was constructed with which is not necessarily the same as the signdedness of the fixed point type itself. [`__mul__` uses .as_value](https://github.com/zyp/amaranth/blob/7c1d8c43b7636a1bc30959ca89d70753254708b9/amaranth/lib/fixed.py#L143) and then constructs the new fixed point type based on the signdedness of the result of the multiplication when it is cast back to a fixed point type:

```python
In [19]: x = fixed.SQ(8, 8)(unsigned(17))

In [20]: x, x.as_value()
Out[20]: ((fixedpoint SQ8.8 unsigned(17)), unsigned(17))

In [21]: y = fixed.Shape.cast(x.as_value())

In [22]: y, y.signed
Out[22]: (fixed.Shape(17, 0, signed=False), False)
```

several of the other arithmetic operators do the same thing and this seems to break proper sign handling.

Reply:
You're not really supposed to be able to make a `fixed.Value` that differs in width/signedness from the underlying `Value`. Expect your first line to raise an error.

Reply:
Ok I've checked my notes, one place where I had a conversion to unsigned from signed was because of doing what you describe intentionally (a hacky attempt at implementing complex numbers in a struct and slicing the underlying unsigned value).

The other two I believe come from [here](https://github.com/amaranth-lang/amaranth/blob/7c1d8c43b7636a1bc30959ca89d70753254708b9/amaranth/lib/fixed.py#L65), and [here](https://github.com/amaranth-lang/amaranth/blob/7c1d8c43b7636a1bc30959ca89d70753254708b9/amaranth/lib/fixed.py#L183) (because `Cat` appears to be treated as unsigned?), eg:

```python
In [53]: x = Signal(fixed.SQ(8, 8))

In [54]: x << 9
Out[54]: (fixedpoint UQ18.0 (cat (const 1'd0) (sig x)))

In [55]: y = x.round(f_width = 9)

In [56]: y, y * y
Out[56]:
((fixedpoint SQ8.9 (cat (const 1'd0) (sig x))),
 (fixedpoint UQ18.18 (* (cat (const 1'd0) (sig x)) (cat (const 1'd0) (sig x)))))
```

Reply:
The signedness rules are available [for all of the operations](https://amaranth-lang.org/docs/amaranth/latest/reference.html#amaranth.hdl.Value).